### PR TITLE
IGNITE-20837 : Partition awareness with node filter

### PR DESCRIPTION
**Pull request by Vladsz83**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionsRequest.java [hunk 1]
```diff
@@ -187,15 +186,6 @@ private static boolean isApplicable(CacheConfiguration<?, ?> ccfg, boolean withC
         if (ccfg.getCacheMode() != CacheMode.PARTITIONED)
             return false;
 
-        IgnitePredicate<?> filter = ccfg.getNodeFilter();
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionsRequest.java:190]: We can't remove this logic. Partition assignments are grouped into `ClientCachePartitionAwarenessGroup`, where caches have the same assignment. Caches with a node filter have different assignment, as explained below.

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionsRequest.java (hunk 1 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionsRequest.java:190]: Why the caches with different assigment are included in single ClientCachePartitionAwarenessGroup? Do we need to split into several separated ClientCachePartitionAwarenessGroup?

Comment (ptupitsyn) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionsRequest.java:190]: Caches with same assignments are included in the same `ClientCachePartitionAwarenessGroup`. If there is a node filter, then we use separate `ClientCachePartitionAwarenessGroup` for such a cache.

I'm not sure if anything needs to be fixed or improved here.

Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionsRequest.java:190]: 'If there is a node filter, then we use separate ClientCachePartitionAwarenessGroup for such a cache'

1) If I'm correct, we group caches with a node filter always in one `ClientCachePartitionAwarenessGroup` with null mapping. And never send the mapping to the client:
 ```
        ClientCachePartitionMapping mapping = null;

        if (isApplicable(cacheDesc.cacheConfiguration(), withCustomMappings))
            mapping = new ClientCachePartitionMapping(assignment);
```
That is why the partition awareness is skipped for such caches. Same for caches with custom affinity.

2) Another interesting issue is that grouping of partitions mapping doesn't seem to work. `equals()` and `hashCode()` of `ClientCachePartitionAwarenessGroup` do not function properly. `ClientCachePartitionAwarenessGroup` checks `ClientCachePartitionMapping` which checks its `HashMap<UUID, Set<Integer>> partitionMap`. The problem is that the set is actually `ImmutableIntSet`. It has no `equals()` and `hashCode()`. Same as the involved `BitSetIntSet`. 

Could you check the new tests?

**Review by ptupitsyn — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionsRequest.java (hunk 1 above)
Comment (ptupitsyn) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionsRequest.java:190]: Approved, thank you.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionAwarenessGroup.java [hunk 2]
```diff
@@ -42,7 +42,7 @@ class ClientCachePartitionAwarenessGroup {
     private final boolean dfltAffinity;
 
     /** Descriptor of the associated caches. */
-    private final Map<Integer, CacheConfiguration<?, ?>> cacheCfgs = new HashMap<>();
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionAwarenessGroup.java:45]: Redundant change

File: modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAbstractPartitionAwarenessTest.java [hunk 3]
```diff
@@ -389,4 +399,14 @@ public boolean isClosed() {
             return cfg.getAddresses().toString();
         }
     }
+
+    /**
+     * Excludes node if its consistent id ends with 'Test1'.
+     */
+    protected static final class ConsistentIdNodeFilter implements IgnitePredicate<ClusterNode> {
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAbstractPartitionAwarenessTest.java:406]: There are three almost identical classes. Can we use just and and pass consistent id suffix as constructor parameter?

File: modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientPartitionAwarenessStableTopologyTest.java [hunk 4]
```diff
@@ -409,14 +600,29 @@ private void testNotApplicableCache(String cacheName) {
...
+        Function<Integer, Object> keyFactory,
+        boolean partitionsRequestExpected
+    ) throws Exception {
         ClientCache<Object, Object> clientCache = client.cache(cacheName);
         IgniteInternalCache<Object, Object> igniteCache = grid(0).context().cache().cache(cacheName);
 
         clientCache.put(keyFactory.apply(0), 0);
 
-        TestTcpClientChannel opCh = affinityChannel(keyFactory.apply(0), igniteCache);
+        TestTcpClientChannel opCh = affinityChannel(0, igniteCache);
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientPartitionAwarenessStableTopologyTest.java:621]: `keyFactory.apply(0)` - correct implementation, why it was changed?

Comment (Vladsz83) [reply modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientPartitionAwarenessStableTopologyTest.java:621]: Hm. Yes. Confused with a key->value factory

File: modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAbstractPartitionAwarenessTest.java (hunk 3 above)
Comment (Vladsz83) [reply modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAbstractPartitionAwarenessTest.java:406]: We check filters equality for a cache group only by their classes. A parametrized filter is a "bad example" for a user.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionAwarenessGroup.java (hunk 2 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionAwarenessGroup.java]: Yes. Left atfer some research.

**Review by alex-plekhanov — APPROVED** [review]

**Pull request closed without merging**